File: DataBindingBasicJava/app/src/main/java/com/websarva/wings/android/databindingbasicjava/MainActivity.java
package com.websarva.wings.android.databindingbasicjava;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProvider;

import android.os.Bundle;
import android.view.View;

import com.websarva.wings.android.databindingbasicjava.databinding.ActivityMainBinding;
import com.websarva.wings.android.databindingbasicjava.ui.MainViewModel;

public class MainActivity extends AppCompatActivity {
	MainViewModel _mainViewModel;
	ActivityMainBinding _activityMainBinding;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
//		_activityMainBinding = ActivityMainBinding.inflate(getLayoutInflater());
//		View contentView = _activityMainBinding.getRoot();
//		setContentView(contentView);
		_activityMainBinding = DataBindingUtil.setContentView(MainActivity.this, R.layout.activity_main);
		ViewModelProvider provider = new ViewModelProvider(MainActivity.this);
		_mainViewModel = provider.get(MainViewModel.class);

		_activityMainBinding.setRandNum(_mainViewModel.getRandNumStr());

		_activityMainBinding.btnRand.setOnClickListener(new ButtonClickListener());
	}

	private class ButtonClickListener implements View.OnClickListener {
		@Override
		public void onClick(View v) {
			_mainViewModel.generateNewRandNum();
			_activityMainBinding.setRandNum(_mainViewModel.getRandNumStr());
		}
	}
}

File: DataBindingBasicKotlin/app/src/main/java/com/websarva/wings/android/databindingbasickotlin/MainActivity.kt
package com.websarva.wings.android.databindingbasickotlin

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import androidx.activity.viewModels
import androidx.databinding.DataBindingUtil
import com.websarva.wings.android.databindingbasickotlin.databinding.ActivityMainBinding
import com.websarva.wings.android.databindingbasickotlin.ui.MainViewModel

class MainActivity : AppCompatActivity() {
	private val _mainViewModel by viewModels<MainViewModel>()
	private lateinit var _activityMainBinding: ActivityMainBinding

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		_activityMainBinding = DataBindingUtil.setContentView(this@MainActivity, R.layout.activity_main)
		_activityMainBinding.randNum = _mainViewModel.getRandNumStr()

		_activityMainBinding.btnRand.setOnClickListener(ButtonClickListener())

	}

	private inner class ButtonClickListener : View.OnClickListener {
		override fun onClick(view: View?) {
			_mainViewModel.generateNewRandNum()
			_activityMainBinding.randNum = _mainViewModel.getRandNumStr()
		}
	}
}
